### ISSUE TITLE
📋 TASK 2: TUÂN - NAIVE BAYES IMPLEMENTATION (DEADLINE: 30/4)

### ISSUE THREAD

**Issue by Minh08012005** [Minh08012005/Heart-Disease-Prediction-ML-Group-6#2]
### TASK-02: Implement Gaussian Naive Bayes từ đầu

**Assignee:** @Tuân  
**Deadline:** 30/4/2026 (7 ngày)  
**Priority:** HIGH 🔴  
**Labels:** algorithm, naive-bayes, from-scratch  
**Status:** To Do  

**🎯 MỤC TIÊU:**
Có class NaiveBayes hoạt động được, predict được trên data đơn giản.

**📝 CÔNG VIỆC CỤ THỂ:**

**Ngày 1-2 (23-24/4): Nghiên cứu toán học**
1. Hiểu Bayes Theorem: P(A|B) = P(B|A) * P(A) / P(B)
2. Hiểu Gaussian Distribution PDF:

P(x|μ,σ²) = 1/√(2πσ²) * exp(-(x-μ)²/(2σ²))

3. Hiểu Naive Bayes assumption: Features độc lập

Ngày 3-4 (25-26/4): Code structure**
```python
# src/models/naive_bayes.py
import numpy as np

class NaiveBayes:
 def __init__(self):
     self.priors = {}      # P(y) cho mỗi class
     self.means = {}       # mean của mỗi feature cho mỗi class
     self.variances = {}   # variance của mỗi feature cho mỗi class
 
 def fit(self, X, y):
     # 1. Tính priors: P(y) = count(y) / n_samples
     # 2. Tính means & variances cho từng feature & class
     pass
 
 def _gaussian_pdf(self, x, mean, var):
     # Tính Gaussian Probability Density Function
     return (1 / np.sqrt(2 * np.pi * var)) * np.exp(-(x - mean)**2 / (2 * var))
 
 def predict(self, X):
     # Với mỗi sample, tính posterior cho mỗi class
     # Chọn class có posterior lớn nhất
     pass
````

__Ngày 5-7 (27-30/4): Testing & Debug__

1. Test với data đơn giản:

   ```python
   X = [[1, 2], [2, 3], [3, 4], [4, 5]]
   y = [0, 0, 1, 1]
   model = NaiveBayes()
   model.fit(X, y)
   predictions = model.predict([[2.5, 3.5]])
   ```

2. Test với dataset thật (sau khi Minh xử lý xong)

__✅ DELIVERABLES (30/4):__

- `src/models/naive_bayes.py` hoàn chỉnh
- Test script chứng minh hoạt động
- Accuracy > 70% trên test set đơn giản

__🔗 TÀI LIỆU THAM KHẢO:__

- [Bayes Theorem Explained](https://www.youtube.com/watch?v=HZGCoVF3YvM)
- [Gaussian Naive Bayes Math](https://www.geeksforgeeks.org/naive-bayes-classifiers/)
- [Code along tutorial](https://github.com/...)

```
```

**Comment by vantuann205** [comment]
Finished

**Comment by Minh08012005** [comment]
Yêu cầu fix và test code Naive Bayes
Code thuật toán ổn, nhưng cần làm các bước sau:

__Bước 1: Fix 2 lỗi nhỏ__

- Dòng 12: Đổi `self.classes = None` → `self.classes = np.array([])`

- Thêm đầu hàm predict và predict_proba:

  ```python
  if len(self.classes) == 0:
      raise ValueError("Model chưa được fit! Hãy gọi fit() trước.")
  ```

__Bước 2: Pull code mới từ main__

```bash
git checkout feature/naive-bayes
git pull origin main
```

(Sẽ có utils.py mới với đầy đủ hàm đánh giá)

__Bước 3: Chạy test với data thật__ Copy lệnh này vào terminal:

```bash
python -c "
from src.models.naive_bayes import NaiveBayes
from src.utils import train_test_split, accuracy_score, confusion_matrix, precision_score, recall_score, f1_score
import pandas as pd
import numpy as np

df = pd.read_csv('data/heart_preprocessed.csv')
X = df.drop('HeartDisease', axis=1).values
y = df['HeartDisease'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = NaiveBayes()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Precision:', precision_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('F1:', f1_score(y_test, y_pred))
print('Confusion Matrix:')
print(confusion_matrix(y_test, y_pred))
"
```

__Bước 4: Push lại code__

@vantuann205  xử lý tiếp nhé ô.
